File: ports/dotnet/src/RussianDocs.DocumentProcessing/Preprocess/Yolo.cs
using RussianDocs.DocumentProcessing.Imaging;
using RussianDocs.DocumentProcessing.Models;
using RussianDocs.DocumentProcessing.Tensors;

namespace RussianDocs.DocumentProcessing.Preprocess;

/// <summary>
/// The detector input: pad, letterbox to the declared square, hand over as uint8 NHWC.
/// </summary>
public sealed class Yolo : IPreprocessor
{
    /// <summary>
    /// The letterbox fill. Grey rather than black because that is what the reference uses, and the
    /// value reaches the model — a different fill shifts every score slightly.
    /// </summary>
    private static readonly byte[] Fill = [114, 114, 114];

    private readonly int _width;
    private readonly int _height;
    private readonly IReadOnlyList<int>? _paddingSize;
    private readonly IReadOnlyList<int>? _paddingColour;

    public Yolo(ModelInput input)
    {
        List<int> shape = input.Shape
            ?? throw new InvalidDataException("preprocess: YOLO input has no Shape");
        if (shape.Count < 2)
        {
            throw new InvalidDataException(
                $"preprocess: YOLO Shape needs at least 2 entries, got {shape.Count}");
        }
        _height = shape[0];
        _width = shape[1];
        _paddingSize = input.PaddingSize;
        _paddingColour = input.PaddingColor;
    }

    public (NdArray Tensor, Meta Meta) Apply(Image image)
    {
        (Image padded, int[] extra) = Padding.Pad(image, _paddingSize, _paddingColour);
        using (padded)
        {
            int paddedH = padded.Height, paddedW = padded.Width;
            (Image boxed, double ratio, double[] padLetter) = Letterbox(padded, _height, _width);
            using (boxed)
            {
                NdArray pixels = boxed.ToArray();
                var batched = new NdArray(pixels.Data,
                    [1, boxed.Height, boxed.Width, boxed.Channels], Dtype.UInt8, 1);
                return (batched, new Meta
                {
                    Ratio = ratio,
                    PadExtra = extra,
                    PadLetter = padLetter,
                    PaddedH = paddedH,
                    PaddedW = paddedW,
                    OrigH = image.Height,
                    OrigW = image.Width,
                });
            }
        }
    }

    /// <summary>
    /// Scales to fit and pads the remainder — the standard YOLO letterbox.
    ///
    /// <para>
    /// **The asymmetric <c>±0.1</c> is not decoration.** With an odd amount of padding to distribute,
    /// <c>round(dh - 0.1)</c> and <c>round(dh + 0.1)</c> put the extra row at the BOTTOM and the
    /// extra column at the RIGHT. A "clean" implementation that halves the padding evenly shifts
    /// every returned box by a pixel, and the failure surfaces as a coordinate mismatch with no
    /// obvious source.
    /// </para>
    ///
    /// <para>
    /// Every rounding here is half-to-even, matching <c>np.round</c>. In .NET that is the default;
    /// the calls are explicit anyway so the intent survives a reader who does not know that.
    /// </para>
    ///
    /// <para>
    /// The resize is SKIPPED when the size already matches, rather than run with a ratio of 1.0 —
    /// running the interpolator needlessly can perturb pixels.
    /// </para>
    /// </summary>
    private static (Image Boxed, double Ratio, double[] PadLetter) Letterbox(
        Image src, int targetH, int targetW)
    {
        int h = src.Height, w = src.Width;

        double ratio = Math.Min((double)targetH / h, (double)targetW / w);
        int newW = PyNum.RoundHalfEvenToInt(w * ratio);
        int newH = PyNum.RoundHalfEvenToInt(h * ratio);

        double dw = (targetW - newW) / 2.0;
        double dh = (targetH - newH) / 2.0;

        Image scaled = w != newW || h != newH
            ? Io.Resize(src, newW, newH, Interpolation.Linear)
            : src.Clone();

        using (scaled)
        {
            int top = PyNum.RoundHalfEvenToInt(dh - 0.1);
            int bottom = PyNum.RoundHalfEvenToInt(dh + 0.1);
            int left = PyNum.RoundHalfEvenToInt(dw - 0.1);
            int right = PyNum.RoundHalfEvenToInt(dw + 0.1);

            Image boxed = Io.CopyMakeBorderConstant(scaled, top, bottom, left, right,
                Fill[0], Fill[1], Fill[2]);
            return (boxed, ratio, [dw, dh]);
        }
    }
}

File: ports/java/docproc/src/main/kotlin/net/russiandocs/docproc/preprocess/Yolo.kt
package net.russiandocs.docproc.preprocess

import net.russiandocs.docproc.imaging.Image
import net.russiandocs.docproc.imaging.Interpolation
import net.russiandocs.docproc.imaging.Io
import net.russiandocs.docproc.models.ModelInput
import net.russiandocs.docproc.tensors.Dtype
import net.russiandocs.docproc.tensors.NdArray
import net.russiandocs.docproc.tensors.PyNum
import kotlin.math.min

/** The detector input: pad, letterbox to the declared square, hand over as uint8 NHWC. */
public class Yolo(input: ModelInput) : Preprocessor {

    private val height: Int
    private val width: Int
    private val paddingSize: List<Int>? = input.paddingSize
    private val paddingColour: List<Int>? = input.paddingColor

    init {
        val shape = input.shape
            ?: throw IllegalArgumentException("preprocess: YOLO input has no Shape")
        require(shape.size >= 2) {
            "preprocess: YOLO Shape needs at least 2 entries, got ${shape.size}"
        }
        height = shape[0]
        width = shape[1]
    }

    override fun apply(image: Image): Pair<NdArray, Meta> {
        val (padded, extra) = Padding.pad(image, paddingSize, paddingColour)
        padded.use {
            val paddedH = padded.height
            val paddedW = padded.width
            val (boxed, ratio, padLetter) = letterbox(padded, height, width)
            boxed.use {
                val pixels = boxed.toArray()
                val batched = NdArray(
                    pixels.data,
                    intArrayOf(1, boxed.height, boxed.width, boxed.channels),
                    Dtype.UINT8,
                    1,
                )
                return batched to Meta(
                    ratio = ratio,
                    padExtra = extra,
                    padLetter = padLetter,
                    paddedH = paddedH,
                    paddedW = paddedW,
                    origH = image.height,
                    origW = image.width,
                )
            }
        }
    }

    /**
     * Scales to fit and pads the remainder — the standard YOLO letterbox.
     *
     * **The asymmetric ±0.1 is not decoration.** With an odd amount of padding to distribute,
     * `round(dh - 0.1)` and `round(dh + 0.1)` put the extra row at the BOTTOM and the extra column at the
     * RIGHT. A "clean" implementation that halves the padding evenly shifts every returned box by a pixel,
     * and the failure surfaces as a coordinate mismatch with no obvious source.
     *
     * Every rounding here is half-to-even, matching `np.round`. On the JVM that means `Math.rint` — NOT
     * `Math.round` or `roundToInt`, both of which are half-up.
     *
     * The resize is SKIPPED when the size already matches, rather than run with a ratio of 1.0 — running
     * the interpolator needlessly can perturb pixels.
     */
    private fun letterbox(src: Image, targetH: Int, targetW: Int): Triple<Image, Double, DoubleArray> {
        val h = src.height
        val w = src.width

        val ratio = min(targetH.toDouble() / h, targetW.toDouble() / w)
        val newW = PyNum.roundHalfEvenToInt(w * ratio)
        val newH = PyNum.roundHalfEvenToInt(h * ratio)

        val dw = (targetW - newW) / 2.0
        val dh = (targetH - newH) / 2.0

        val scaled = if (w != newW || h != newH) {
            Io.resize(src, newW, newH, Interpolation.LINEAR)
        } else {
            src.clone()
        }

        scaled.use {
            val top = PyNum.roundHalfEvenToInt(dh - 0.1)
            val bottom = PyNum.roundHalfEvenToInt(dh + 0.1)
            val left = PyNum.roundHalfEvenToInt(dw - 0.1)
            val right = PyNum.roundHalfEvenToInt(dw + 0.1)

            val boxed = Io.copyMakeBorderConstant(scaled, top, bottom, left, right,
                FILL, FILL, FILL)
            return Triple(boxed, ratio, doubleArrayOf(dw, dh))
        }
    }

    private companion object {
        /**
         * The letterbox fill. Grey rather than black because that is what the reference uses, and the value
         * reaches the model — a different fill shifts every score slightly.
         */
        const val FILL = 114.0
    }
}
